### DISCUSSION TITLE
Fluentd Log Ordering Issue in Multi-Worker

### DISCUSSION THREAD

**Discussion by mikankun1003** [fluent/fluentd#4634] [Q&A]
I'm using Fluentd v1.16.1 in a multi-worker environment:

- Forwarders use `in_tail` to read logs and `out_forward` to transfer them.
- The Aggregator uses `in_forward` to receive logs and `out_file` (with Chunk_keys Tag file Buffer) to write them to files.

We're running a single Aggregator with multiple workers. When the Aggregator encounters an issue, each Forwarder buffers Chunks locally and transfers these local Chunks simultaneously upon recovery.

This leads to the Aggregator temporarily receiving a large volume of logs. As each Worker processes these asynchronously, it can result in logs being written out of order.

Is there a way to guarantee the order of logs? 

Note: It would be ideal if we could sort the log records before writing them to the file.

**Comment by cosmo0920** [comment]
> Is there a way to guarantee the order of logs?

There is no 100% ensuring way when sending chunks with retrying. This is because it also requires that it will block the latter part of logs. This couldn't be supported for the current architecture of Fluentd. 
In these days, there are several products that tolerate to accept out-of-order logs like Prometheus. Prometheus' TSDB had already accepted out-of-order logs.
see: https://promlabs.com/blog/2022/10/05/whats-new-in-prometheus-2-39/
